### PR TITLE
[dynamo] support torch.dtype and torch.device in default class arguments

### PR DESCRIPTION
**Pull request by d4l3k**
Related to #109910

Test plan:

```
pytest test/dynamo/test_misc.py -k torch_device
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1848,7 +1848,7 @@ def __call__(self, tx, value) -> VariableTracker:
             return SourcelessBuilder.wrap_constant_literal(value)
         elif is_builtin_callable(value):
             return BuiltinVariable(value)
-        elif is_allowed(value):
+        elif is_allowed(value) or isinstance(value, (torch.device, torch.dtype)):
             if is_user_defined_allowed(value):
                 self.tx.output.has_user_defined_allowed_in_graph = True
             return TorchVariable(value)
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builder.py:1854]: I don't think `torch.device/torch.dtype` are torch variables. 

They are `ConstantVariable`s

You need to change the definition of `is_literal`

File: test/dynamo/test_misc.py [hunk 2]
```diff
@@ -7511,6 +7511,22 @@ def fn(target):
...
+                device: torch.device = torch.device("cpu"),
+            ) -> None:
+                self.value = torch.tensor(10, dtype=dtype, device=device)
+
+        def fn():
+            return Foo().value
+
+        opt_func = torch._dynamo.optimize("inductor", nopython=True)(fn)
+        res = opt_func()
+        self.assertEqual(res, 10.0)
```
Comment (jon-chuang) [inline test/dynamo/test_misc.py:7528]: You should assert that there is no graph break.

See `CompileCounter`

File: test/dynamo/test_misc.py [hunk 3]
```diff
@@ -7511,6 +7511,22 @@ def fn(target):
...
+                self,
+                dtype: torch.dtype = torch.float,
+                device: torch.device = torch.device("cpu"),
+            ) -> None:
+                self.value = torch.tensor(10, dtype=dtype, device=device)
+
+        def fn():
+            return Foo().value
+
+        opt_func = torch._dynamo.optimize("inductor", nopython=True)(fn)
```
Comment (jon-chuang) [inline test/dynamo/test_misc.py:7526]: Please use "eager" if inductor is not strictly required.

File: test/dynamo/test_misc.py [hunk 4]
```diff
@@ -7511,6 +7511,22 @@ def fn(target):
...
+        class Foo:
+            def __init__(
+                self,
+                dtype: torch.dtype = torch.float,
+                device: torch.device = torch.device("cpu"),
+            ) -> None:
+                self.value = torch.tensor(10, dtype=dtype, device=device)
+
+        def fn():
+            return Foo().value
```
Comment (jon-chuang) [inline test/dynamo/test_misc.py:7524]: Please perform a tensor computation `Foo().value + 1` in order to make sure it compiles

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (d4l3k) [reply torch/_dynamo/variables/builder.py:1854]: The existing tests for torch.device show it as a TorchVariable 

https://github.com/pytorch/pytorch/blob/85dedeab654366400028e83cf34c8daad8a6afb9/test/dynamo/test_misc.py#L7505-L7507

Is torch.device a TorchVariable but torch.dtype should be a constant?

File: test/dynamo/test_misc.py (hunk 4 above)
Comment (d4l3k) [reply test/dynamo/test_misc.py:7524]: done

File: test/dynamo/test_misc.py (hunk 3 above)
Comment (d4l3k) [reply test/dynamo/test_misc.py]: done

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/builder.py:1854]: Hmm that's weird, as variablebuilder seems to wrap it as a constantvariable. I guess both are technically fine, but changing is_literal can get rid of the special-casing here.

Comment (jon-chuang) [reply torch/_dynamo/variables/builder.py:1854]: https://github.com/pytorch/pytorch/blob/919c9b713e65aa1489123259a0ce7e79322d4d9d/torch/_dynamo/variables/builder.py#L302

Comment (yanboliang) [reply torch/_dynamo/variables/builder.py:1854]: ```torch.device/torch.dtype``` should be ```ConstantVariable```.

File: torch/_dynamo/variables/builder.py [hunk 5]
```diff
@@ -1848,7 +1848,7 @@ def __call__(self, tx, value) -> VariableTracker:
             return SourcelessBuilder.wrap_constant_literal(value)
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builder.py:1848]: this is the code path that I prefer device and dtype to take

Comment (drisspg) [reply torch/_dynamo/variables/builder.py:1848]: https://github.com/pytorch/pytorch/pull/112284

Comment (jon-chuang) [reply torch/_dynamo/variables/builder.py:1848]: @drisspg since this PR includes a test, could we close that in favour of this one?

Comment (drisspg) [reply torch/_dynamo/variables/builder.py:1848]: Forsure, tbh I wasn't able to find any sourceless builder tests and was curious how where to add one, will close

**Comment by jon-chuang** [comment]
> What's the path forward here? Sounds like I shouldn't change this to use ConstantVariable?

Actually, I think that one is turning out fine -  we also get rid of some guards special-casing

**Comment by d4l3k** [comment]
@jon-chuang got it -- feel free to land that one if it's working out

**Comment by jon-chuang** [comment]
Landed the sister PR with less special casing

**Pull request closed without merging**